app/services/VirtualMachineService.ts

```diff
@@ -0,0 +1,584 @@
+      }
+
+      const listOfVM = await this.getListofUsedVMID();
+
+      // Assign new vmid depending on the current vm
+      const vmPrisma = await this.assignNewVMID(
+        exerciseId,
+        user,
+        newNode,
+        listOfVM
+      );
+
+      const newId = String(vmPrisma.vmId);
+      await this.cloneTemplate(exerciseId, exerciseNode, newId);
+
+      if (newNode !== exerciseNode) {
+        await this.migrateTemplate(newId, exerciseNode, newNode);
+      }
+
+      await this.startVM(newId, newNode);
+
+      return vmPrisma;
+    } catch (e) {
+      return e;
+    }
+  },
+  /**
+   * Checks if the user already has a VM running
+   * and throw error if exists
+   *
+   * @param {string} user user currently logged in
+   * @throw virtual machine conflict exception
+   *
+   * @return {Promise<void | VirtualMachineConflictException>}
+   * */
+  async checkRunningVM(
+    user: string
+  ): Promise<void | VirtualMachineConflictException> {
+    const VM = await prisma.vM.findFirst({
+      where: {
+        user: user,
+      },
+    });
+    // Throw exception if vm is in running state
+    if (VM !== null && VM.status === "running") {
+      throw new VirtualMachineConflictException();
+    }
+    return;
+  },
+
+  /**
+   * checks the next lowest number to use for newId
+   * Min and Max are determined from global variable
+   *
+   * @param {string} exerciseId vmid for the exercise
+   * @param {string} user userId of current user
+   * @param {string} newLoad name of the node
+   * @param {string []} listOfVM list of vm to avoid when assigning vmid
+   * @return {string} newId to use for cloning
+   */
+  async assignNewVMID(
+    exerciseId: string,
+    user: string,
+    newLoad: string,
+    listOfVM: string[]
+  ): Promise<VM> {
+    let newId: number = minVMID;
+
+    // Find the list number of VMID without range
+    for (let i = minVMID; i <= maxVMID; i++) {
+      const found = listOfVM.find((element) => element === String(i));
+      if (found === undefined) {
+        newId = i;
+        break;
+      }
+    }
+
+    // Get the end time of the vmid
+    const metadata: Object = ExerciseService.getMetaData(exerciseId);
+    const endTime: Date = this.getVMEndTime(metadata["timelimit"]);
+
+    const stringId = String(newId);
+
+    // Check if it already exist
+    const vmUser = await prisma.vM.findFirst({
+      where: {
+        user: user,
+        exerciseId: exerciseId,
+      },
+    });
+    if (vmUser === null) {
+      // Create the to vmid so that it will be there
+      return await prisma.vM.create({
+        data: {
+          user: user,
+          vmId: stringId,
+          node: newLoad,
+          exerciseId: exerciseId,
+          ip: "",
+          port: "",
+          timeLimit: metadata["timelimit"],
+          timeEnd: endTime,
+          status: "running",
+        },
+      });
+    } else {
+      return await prisma.vM.update({
```

Comment:
Test if you can make the prisma.vm lowercase or uppercase. If you can make it a consistent casing, do that.

Reply:
Doing it all uppercase works. Will change all the vm prisma to uppercases